libcudacxx/include/cuda/__barrier/barrier_thread_scope_block.h

```diff
@@ -0,0 +1,466 @@
+        asm volatile("mbarrier.inval.shared.b64 [%0];" ::"r"(static_cast<_CUDA_VSTD::uint32_t>(
+          __cvta_generic_to_shared(&__barrier)))
+                     : "memory");
+      }))
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI friend void init(
+    barrier* __b, _CUDA_VSTD::ptrdiff_t __expected, _CUDA_VSTD::__empty_completion = _CUDA_VSTD::__empty_completion())
+  {
+    NV_DISPATCH_TARGET(
+      NV_PROVIDES_SM_90,
+      (
+        if (__isShared(&__b->__barrier)) {
+          asm volatile("mbarrier.init.shared.b64 [%0], %1;" ::"r"(
+                         static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(&__b->__barrier))),
+                       "r"(static_cast<_CUDA_VSTD::uint32_t>(__expected))
+                       : "memory");
+        } else if (__isClusterShared(&__b->__barrier)) { __trap(); } else {
+          new (&__b->__barrier) __barrier_base(__expected);
+        }),
+      NV_PROVIDES_SM_80,
+      (
+        if (__isShared(&__b->__barrier)) {
+          asm volatile("mbarrier.init.shared.b64 [%0], %1;" ::"r"(
+                         static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(&__b->__barrier))),
+                       "r"(static_cast<_CUDA_VSTD::uint32_t>(__expected))
+                       : "memory");
+        } else { new (&__b->__barrier) __barrier_base(__expected); }),
+      NV_ANY_TARGET,
+      (new (&__b->__barrier) __barrier_base(__expected);))
+  }
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI arrival_token arrive(_CUDA_VSTD::ptrdiff_t __update = 1)
+  {
+    _LIBCUDACXX_DEBUG_ASSERT(__update >= 0, "Arrival count update must be non-negative.");
+    arrival_token __token = {};
+    NV_DISPATCH_TARGET(
+      NV_PROVIDES_SM_90,
+      (
+        if (!__isClusterShared(&__barrier)) { return __barrier.arrive(__update); } else if (!__isShared(&__barrier)) {
+          __trap();
+        }
+        // Cannot use cuda::device::barrier_native_handle here, as it is
+        // only defined for block-scope barriers. This barrier may be a
+        // non-block scoped barrier.
+        auto __bh = reinterpret_cast<_CUDA_VSTD::uint64_t*>(&__barrier);
+        __token   = _CUDA_VPTX::mbarrier_arrive(__bh, __update);),
+      NV_PROVIDES_SM_80,
+      (
+        if (!__isShared(&__barrier)) {
+          return __barrier.arrive(__update);
+        } auto __bh = reinterpret_cast<_CUDA_VSTD::uint64_t*>(&__barrier);
+        // Need 2 instructions, can't finish barrier with arrive > 1
+        if (__update > 1) { _CUDA_VPTX::mbarrier_arrive_no_complete(__bh, __update - 1); } __token =
+          _CUDA_VPTX::mbarrier_arrive(__bh);),
+      NV_PROVIDES_SM_70,
+      (
+        if (!__isShared(&__barrier)) { return __barrier.arrive(__update); }
+
+        unsigned int __mask    = __activemask();
+        unsigned int __activeA = __match_any_sync(__mask, __update);
+        unsigned int __activeB = __match_any_sync(__mask, reinterpret_cast<_CUDA_VSTD::uintptr_t>(&__barrier));
```

Comment:
@wmaxey I believe this is kind of a fix. because pre SM70 there is no `__match_any_sync`